finesse/hardware/dummy_opus.py

```diff
@@ -0,0 +1,168 @@
+        """Log all state transitions."""
+        logging.info(f"Current state: {target.name}")
+
+
+class DummyOPUSInterface(OPUSInterfaceBase):
+    """A mock version of the OPUS API for testing purposes."""
+
+    _request_status = Signal()
+    _request_command = Signal(str)
+
+    def __init__(self, measure_duration: float = 1.0) -> None:
+        """Create a new DummyOPUSInterface.
+
+        Args:
+            measure_duration: How long a single measurement takes (seconds)
+        """
+        super().__init__()
+
+        self.last_error = OPUSError.NO_ERROR
+        """The last error which occurred."""
+        self.state_machine = OPUSStateMachine()
+        """An object representing the internal state of the device."""
+
+        self.measure_timer = QTimer(self)
+        """Timer signalling the end of a measurement."""
+        self.measure_timer.setInterval(round(measure_duration * 1000))
+        self.measure_timer.setSingleShot(True)
+        self.measure_timer.timeout.connect(self.finish_measuring)  # type: ignore
+
+    def __del__(self) -> None:
+        """Stop the background timer."""
+        self.measure_timer.stop()
+
+    def _send_response(self, type: str) -> None:
+        """Send a message signalling that a response was received."""
+        state = self.state_machine.current_state
+
+        pub.sendMessage(
+            f"opus.{type}.response",
+            url="https://example.com",
+            status=state.value,
+            text=state.name,
+            error=self.last_error.to_tuple(),
+        )
+
+    def request_status(self) -> None:
+        """Request the device's current status."""
+        if self.state_machine.current_state == OPUSStateMachine.idle:
+            self.last_error = OPUSError.NOT_CONNECTED
+
+        self._send_response("status")
+
+    def request_command(self, command: str) -> None:
+        """Execute the specified command on the device."""
+        self.last_error = OPUSError.NO_ERROR
+
+        if command == "cancel":
+            try:
+                self.state_machine.cancel()
+                self.measure_timer.stop()
+            except TransitionNotAllowed:
+                self.last_error = OPUSError.NOT_RUNNING
+        elif command == "stop":
+            try:
+                self.state_machine.stop()
+                self.measure_timer.stop()
+            except TransitionNotAllowed:
+                # TODO: This currently won't work if it actually *is* finishing
+                self.last_error = OPUSError.NOT_RUNNING_OR_FINISHING
+        elif command == "start":
+            try:
+                self.state_machine.start_measuring()
+                self.measure_timer.start()
+            except TransitionNotAllowed:
+                self.last_error = OPUSError.NOT_CONNECTED
+        elif command == "connect":
+            try:
+                self.state_machine.connect()
+            except TransitionNotAllowed:
+                self.last_error = OPUSError.NOT_IDLE
```

Comment:
This is a dummy device and we should not invest more time than necessary here, but I feel this could be abstracted somehow. Repeating the same block of code 4 times with just small differences seems a waste. At the very least, I'd try to use the new [`match` syntax](https://docs.python.org/3.10/whatsnew/3.10.html#pep-634-structural-pattern-matching), available from Python `3.10` (are we using Python 3.10?`)

Reply:
This section is a bit boilerplate-y :disappointed:, though the branches are subtly different from one another.

One option would be to have each command represented as a member function (i.e. you'd have `start()`, `stop()` etc.) and then we could just call the corresponding function for the command. Then we could probably have a decorator which returns a specified error when `TransitionNotAllowed` is raised. What do you think?

I've never used the `match` syntax, but it looks super useful! The minimum requirement for FINESSE is currently 3.9, but I was thinking about bumping it to 3.10. Do you think it'd be worth doing a separate PR for this?

Reply:
Ping @dalonsoa ?

Reply:
Ok cool.

What do you think about this bit?

> One option would be to have each command represented as a member function (i.e. you'd have `start()`, `stop()` etc.) and then we could just call the corresponding function for the command. Then we could probably have a decorator which returns a specified error when `TransitionNotAllowed` is raised. What do you think?

I do have some work based on top of this now, so it might be easier to merge this as is and do the refactoring in my next PR.

Reply:
Separating things into different functions won't be a bad idea, although the decorator part feels a bit overkilling. What about having the following function (appropriately formatted and typed):

```python
def execute(self, *args, error: OPUSError):
    try:
        for fun in args:
            fun()
    except TransitionNotAllowed:
            self.last_error = error
```

And then call it:

```python
        if command == "cancel":
            self.execute(self.state_machine.cancel, self.measure_timer.stop, error = OPUSError.NOT_RUNNING)
        elif command == "stop":
            self.execute(self.state_machine.stop, self.measure_timer.stop, error =  OPUSError.NOT_RUNNING_OR_FINISHING)
        ...
```

Or something along these lines?

Reply:
Good idea.

I've done something similar. I realised it would probably be simpler if `OPUSStateMachine` was responsible for starting/stopping the `QTimer`, which removes one of the differences between commands. I've tried using a `dict` for each of the different error codes instead.

Let me know what you think!